### PR TITLE
HDDS-5959. Handles bucket layout validation logic in ofs/o3fs client.

### PR DESCRIPTION
**Pull request by JyotinderSingh**
## What changes were proposed in this pull request?

Throw an exception if a bucket accessed via OFS is in OBJECT_STORE layout. Allowed layouts are FILE_SYSTEM_OPTIMIZED and LEGACY.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5959

## How was this patch tested?

Related integration tests.

### CODE REVIEW

**Comment by rakeshadr** [comment]
Thanks @JyotinderSingh for taking this ahead!

I've gone through the test case failures quickly and all the contract test cases expect FS semantics. How about setting

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java [hunk 1]
```diff
@@ -153,6 +153,17 @@ public BasicOzoneClientAdapterImpl(String omHost, int omPort,
     objectStore = ozoneClient.getObjectStore();
     this.volume = objectStore.getVolume(volumeStr);
     this.bucket = volume.getBucket(bucketStr);
+
+    // Check if bucket layout is valid, OFS buckets cannot be in
+    // OBJECT_STORE layout
+    BucketLayout bucketLayout = bucket.getBucketLayout();
+    if (bucketLayout.equals(BucketLayout.OBJECT_STORE)) {
+      throw new IllegalArgumentException(bucketLayout + " does not support" +
```
Comment (rakeshadr) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:161]: @JyotinderSingh Please modify the message with bucket name, something like:

  ```
    throw new IllegalArgumentException("Bucket: " + bucket.getName() +" is with layout: " + bucketLayout + ", which does not support" +
          " file system semantics. Bucket Layout must be " +
          BucketLayout.FILE_SYSTEM_OPTIMIZED + " or "
          + BucketLayout.LEGACY + ".");
```

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java [hunk 2]
```diff
@@ -226,6 +226,17 @@ private OzoneBucket getBucket(String volumeStr, String bucketStr,
     OzoneBucket bucket;
     try {
       bucket = proxy.getBucketDetails(volumeStr, bucketStr);
+
+      // Check if bucket layout is valid, OFS buckets cannot be in
+      // OBJECT_STORE layout
+      BucketLayout bucketLayout = bucket.getBucketLayout();
+      if (bucketLayout.equals(BucketLayout.OBJECT_STORE)) {
+        throw new IllegalArgumentException(bucketLayout + " does not support" +
```
Comment (rakeshadr) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:234]: please modify message like above.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java (hunk 1 above)
Comment (mukul1987) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:161]: I will suggest to place this in a common class like FSUtils.java.

**Comment by JyotinderSingh** [comment]
Thanks for the feedback @rakeshadr @mukul1987 . I have made the relevant changes.

However currently the FS related acceptance tests are still failing due to default `OBJECT_STORE` layout behaviour of buckets configured in [ozone-default.xml#2985](https://github.com/apache/ozone/blob/c2db3776773c9a80e44a4c1dbd58d38ff6a41f9f/hadoop-hdds/common/src/main/resources/ozone-default.xml#L2985). I have tried overriding that by adding `OZONE-SITE.XML_ozone.default.bucket.layout=LEGACY` to `docker-config` and `common-config` files of the related `dist/src/main/compose/{test}/*` directories. But for some reason when the corresponding docker compose cluster is started this config not picked up - and the default config value of `OBJECT_STORE` is still being used. Any clues why this might be happening?

**Comment by rakeshadr** [comment]
> Thanks for the feedback @rakeshadr @mukul1987 . I have made the relevant changes - please let me know if they are fine.
> 
> However currently the FS related acceptance tests are still failing due to default `OBJECT_STORE` layout behaviour of buckets configured in [ozone-default.xml#2985](https://github.com/apache/ozone/blob/c2db3776773c9a80e44a4c1dbd58d38ff6a41f9f/hadoop-hdds/common/src/main/resources/ozone-default.xml#L2985). I have tried overriding that by adding `OZONE-SITE.XML_ozone.default.bucket.layout=LEGACY` to `docker-config` and `common-config` files of the related `dist/src/main/compose/{test}/*` directories. But for some reason when the corresponding docker compose cluster is started this config not picked up - and the default config value of `OBJECT_STORE` is still being used. Any clues why this might be happening?
I will explore this part and update to you.
BTW, @sadanand48 do you have clues to fix this case? Thanks for your time!

**Comment by rakeshadr** [comment]
@JyotinderSingh  Please fix TestOzoneFSInputStream case , seems it will pass with those config values.

**Comment by sadanand48** [comment]
> BTW, @sadanand48 do you have clues to fix this case? Thanks for your time!

@rakeshadr , The robot test tries to create bucket via cli through `ozone bucket create `and when looking into the [code](https://github.com/apache/ozone/blob/1b7072cb106429e52ab584e17502562c02eeccd1/hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/CreateBucketHandler.java#L58) i.e in `CreateBucketHandler` we don’t obtain the default bucket type from conf instead it’s taken as OBJECT_STORE . That’s probably why it wasn’t obeying the conf.  I guess the latest change by @JyotinderSingh to explicitly create fso buckets for these tests should work.

**Comment by JyotinderSingh** [comment]
I think most of the tests are now passing. There does seem to be a compatibility test issue though.
the xcompat test is failing on the following test
`xcompat-cluster-1.1.0-client-1.1.0-write :: Write Compatibility`
The test is failing on the following command:
`ozone fs -mkdir o3fs://bucket1.vol1/dir-1.1.0`
This is failing due the the directory being created being in OBS fashion.
```
-mkdir: Bucket: bucket1 has layout: OBJECT_STORE, which does not support file system semantics. Bucket Layout must be FILE_SYSTEM_OPTIMIZED or LEGACY.
```
From my understanding - this doesn’t look like a simple fix - I believe the reason it is failing is that the old 1.1.0 client doesn’t know that the buckets created by the FS need to be in FSO layout - since we added that change in #2730.

I'm not sure how we should handle this.

**Review by rakeshadr — APPROVED** [review]
Thanks @JyotinderSingh for the continuous efforts in refining this.
Thanks @mukul1987 for the reviews.

+1 LGTM, I will merge this shortly.

**Pull request merged**